[JSC] TypedArray construction should scan JSArray in faster way

Constellation opened the pull request:
#### 01b454ac1e6860442583d4b6157857047dc8a616
<pre>
[JSC] TypedArray construction should scan JSArray in faster way
<a href="https://bugs.webkit.org/show_bug.cgi?id=259241">https://bugs.webkit.org/show_bug.cgi?id=259241</a>
rdar://112315641

Reviewed by Michael Saboff.

Use SIMD function to convert JSValue to TypedArray destination types.
By leveraging ARM64 ld2, ld4, we can extract Int32 from JSValue efficiently, and we can store
them to destination.

1. Extract StringImpl::copyCharacters and define WTF::copyElements. It does super fast SIMD copies based on source and destination types.
   It can include widening or narrowing (so source and destination types may be different).
2. Use WTF::copyElements in TypedArray#from and TypedArray constructors.

                                                     ToT                     Patched

typed-array-from-array                        134.4441+-0.4424     ^     13.6653+-0.1270        ^ definitely 9.8383x faster
typed-array-new-from-int32-array               14.5654+-0.0179     ^      2.5779+-0.0212        ^ definitely 5.6501x faster
typed-array-new-from-double-array-to-float64-typed-array
                                                6.0447+-0.0299     ^      3.7598+-0.0156        ^ definitely 1.6077x faster

* JSTests/microbenchmarks/typed-array-new-from-double-array-to-float64-typed-array.js: Added.
(test):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
(JSC::JSGenericTypedArrayView&lt;Adaptor&gt;::setFromArrayLike):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewPrivateFuncFromFast):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::copyElements):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::copyCharacters):

Canonical link: <a href="https://commits.webkit.org/266182@main">https://commits.webkit.org/266182@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8198e3d7826fbbd1bf4596553c6aa69d809015b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12875 "14 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13201 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13530 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14615 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13222 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14989 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13041 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13758 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10897 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15066 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11046 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11648 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18716 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10952 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12121 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15018 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12186 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12279 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10180 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12926 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11540 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3401 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15856 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13297 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1490 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12119 "Built successfully") | | [⏳ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Tests-EWS "Waiting to run tests") | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -843,6 +848,218 @@ inline bool equalIgnoringASCIICase(ASCIILiteral a, ASCIILiteral b)
+        *destination++ = *source++;
+}
+
+inline void copyElements(uint8_t* destination, const uint64_t* source, unsigned length)
```
maybe `restrict` could be used and the non-SIMD while loops would get auto-vectorized?

Jarred-Sumner replied:
I also wonder if the assembly is different/better if size_t is used instead of unsigned

Constellation replied:
`size_t` does not matter since this will be inlined.

Constellation replied:
For restrict, I think it does not affect on the result since SIMD asm path will be used. But anyway, it would be good for non SIMD architectures.

Jarred-Sumner replied:
👍

Constellation replied:
Actually, since now copyElements can be used for non string, size_t would be better as a API (String is always unsigned length, but others may not be true).

msaboff reviewed (commented):
r=me

Constellation committed:
[JSC] TypedArray construction should scan JSArray in faster way
https://bugs.webkit.org/show_bug.cgi?id=259241
rdar://112315641

Reviewed by Michael Saboff.

Use SIMD function to convert JSValue to TypedArray destination types.
By leveraging ARM64 ld2, ld4, we can extract Int32 from JSValue efficiently, and we can store
them to destination.

1. Extract StringImpl::copyCharacters and define WTF::copyElements. It does super fast SIMD copies based on source and destination types.
   It can include widening or narrowing (so source and destination types may be different).
2. Use WTF::copyElements in TypedArray#from and TypedArray constructors.

                                                     ToT                     Patched

typed-array-from-array                        134.4441+-0.4424     ^     13.6653+-0.1270        ^ definitely 9.8383x faster
typed-array-new-from-int32-array               14.5654+-0.0179     ^      2.5779+-0.0212        ^ definitely 5.6501x faster
typed-array-new-from-double-array-to-float64-typed-array
                                                6.0447+-0.0299     ^      3.7598+-0.0156        ^ definitely 1.6077x faster

* JSTests/microbenchmarks/typed-array-new-from-double-array-to-float64-typed-array.js: Added.
(test):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:
(JSC::JSGenericTypedArrayView<Adaptor>::setFromArrayLike):
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewPrivateFuncFromFast):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::copyElements):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::copyCharacters):

Canonical link: https://commits.webkit.org/266182@main
